README.md
# TenTecRebel
Custom firmware for TenTec Rebel 506

## Purpose
This project is designed to replace the factory firmware to add features to the radio. The primary goal is to control as many features as possible through commands sent over the serial port, with the end result being a touch-screen graphical interface while leaving the original hardware controls with their original functionality.

## Features
Some features in the original firmware have been removed or changed because they were unnecessary or unsuitable for the purpose of this project. Many other features have been added to increase the flexibility and functionality of the radio.

### Removed and Changed Features
 * Removed all LCD functionality
 * Removed human-readable data over the serial port
 * Cleaned an organized code for greatly reduced size and improved efficiency
 * Removed tuning step indicator; tuning will be done on the computer now

### Added features
 * Binary serial communications protocol (see below) that includes several different kinds of data and implements a wide range of commands
 * Iambic keyer, or straight key
 * Morse code encoder
 * Morse code decoder

## Computer Interface
Currently the only computer interface available is for testing and debugging. In the future there will be a cross-platform application to act as a front-end for the radio.

## Default Settings
The following are the default values for each setting:

Setting | Value
--------|------
Frequency | 7.03MHZ for 40 meters or 14.06MHz for 20 meters
Filter Bandwidth | 2.5KHz
Tuning Step Size | 100Hz
Keyer Mode | Iambic
Morse Decoder | Enabled
Keyer/Decoder Speed | 20WPM
Decode Threshhold | 700
User Mode | 1 (not used)
Band Scan Mode | Disabled
Band Scan Interval | 1000
Band Scan Width | 5000
Tuning Wraparound | Stop at end

# Serial Protocol
Two-way communication between the radio and a computer is achieved over the USB virtual serial port using a custom protocol.

## Message Format
Each message contains a header byte (0xee), a data type byte, an optional argument, and a footer byte (0xff). At this time, no checksum is utilized. Messages may be either an immediate command or status with no data, may contain a numeric data portion of four bytes, or a variable-length string.

## Output Messages
These message are sent from the radio to the computer:

Message | Number | Data Type | Description
--------|--------|-----------|------------
Text | 0x00 | Variable-length string | Text to display to the user
Transmit Start | 0x01 | None | Radio transmitting
Transmit End | 0x02 | None | Transmission ended (radio receiving)
Voltage In | 0x03 | Integer | 0 to 1023 (raw ADC reading, multiply by 0.01611328125 for actual volts)
Transmit Frequency | 0x04 | Integer | Transmitting frequency in Hz
Receive Frequency | 0x05 | Integer |Receiving (RIT-adjusted) frequency in Hz
Output Power | 0x06 | Integer | RF voltage reading for output power (0 to 1023, meaningful-value calculation forthcoming)
Signal Strength | 0x07 | Integer | Relative signal strength (0 to 1023, meaningful-value calculation forthcoming)
Keyer/Decoder Speed | 0x08 | Integer | Morse code keyer and decoder speed in Words Per Minute (WPM)
Received Character | 0x09 | ASCII Character |Single character received by Morse decoder
Character Sent | 0x0a | None | Character sent by Morse encoder has finished sending
Filter Bandwidth | 0x0b | Integer | Current filter bandwidth selection (0=2.5KHz, 1=1.5KHz, 2=800Hz)
Tuning Step Size | 0x0c | Integer | Current tuning step size selection (0=100Hz, 1=1KHz, 2=10KHz)
User Mode | 0x0d | Integer | Current user mode selected (0, 1, or 2); not used
Keyer Mode | 0x0e | Integer | Keyer mode (0=disabled, 1=straight, 2=iambic, 3=tune)
Initialization Finished | 0x0f | None | Radio initialization has finished and commands may now be sent
Decoder Enable | 0x10 | Boolean | Morse code decoder enabled or disabled (0=off, 1=on)
Decode Threshhold | 0x11 | Integer | Audio level required for a signal to be considered for decoding (0 to 1023)
Band Changed| 0x12 | Integer | Selected band (0=20 meters, 1=40 meters)
Band Scan Mode | 0x13 | Integer | Selected band scan mode (0=off, 1=full band, 2=centered on current frequency, 3=custom)
Band Scan Width | 0x14 | Integer | Scan width for centered scan in Hz
Band Scan Top | 0x15 | Integer | High end for custom band scan in Hz
Band Scan Bottom | 0x16 | Integer | Low end for custom band scan in Hz
Band Scan Interval | 0x17 | Integer | Interval for scan in ms
Band Scan Data | 0x18 | Byte Array | Scan data, divided into 256 bytes
Tuning Wraparound | 0x19 | Integer | 0=stop tuning at band end, 1=wrap to the other end when trying to tune past band end
CW Read Level | 0x1a | Integer | CW read level (0 to 1023)

## Input Messages
These messages are sent from the computer to control the radio:

Message | Number | Data Type | Description
--------|--------|-----------|------------
Set Frequency | 0x80 | Integer | New frequency in Hz (must be valid for the selected band)
Tune Up | 0x81 | None | Tune up a step (based on currently selected tuning step)
Tune Down | 0x82 | None | Tune down a step (based on currently selected tuning step)
Set Keyer/Decoder Speed | 0x83 | Integer | Set keyer and decoder speed in WPM
Send Character | 0x84 | ASCII Character | Send a character using the Morse encoder (only available when the iambic keyer is enabled)
Set Filter Bandwidth | 0x85 | Integer | Set filter width (0=2.5KHz, 1=1.5KHz, 2=800Hz)
Set Tuning Step Size | 0x86 | Integer | Set tuning step size (0=100Hz, 1=1KHz, 2=10KHz)
Set User Mode | 0x87 | Integer | Set user mode (0, 1, or 2); not used
Set Keyer Mode | 0x88 | Integer | Set keyer mode: 0=disabled (can't transmit), 1=straight, 2=iambic, 3=tune (continually transmit)
Set Morse Decoder | 0x89 | Boolean | Enable or disable the Morse decoder (0=disabled, 1=enabled)
Set Morse Decoding Threshhold | 0x8a | Integer | Set audio level required for a signal to be considered for decoding (0 to 1023)
Set Band Scan Mode | 0x8b | Integer | Set band scan mode (0=off, 1=full band, 2=centered on current frequency, 3=custom)
Set Band Scan Width | 0x8c | Integer | Set band scan width (for centered mode) in Hz
Set Band Scan Top | 0x8d | Integer | Set band scan highest frequency in Hz
Set Band Scan Bottom | 0x8e | Integer | Set band scan lowest frequency in Hz
Set Band Scan Interval | 0x8f | Integer | Set interval between band scan sweeps in ms
Set Wraparound Mode | 0x90 | Integer | Set wraparound mode: 0=stop tuning at band end, 1=wrap to the other end when trying to tune past band end
Get All Data | 0x91 | None | Request a dump of all radio data


CHANGELOG.txt
v1.0.18
	CHG Changed rotary encoder code for higher resolution based on code by KD0UTH
v1.0.17 - 2015-01-02
	ADD Added band scope
	ADD Added tuning wraparound at band end
	CHG Renamed some more defines to make more sense
	ADD Added "tune" mode to keyer; changed keyer from iambic on/off to multi-mode
	ADD Added "update" command; still needs to be finished
	DEL Commented out parts of serialDump so it always sends all data
v1.0.16 - 2016-01-01
	CHG Set initial values in serialDump() to ensure that something is sent the first time it is called
	CHG Rearranged some stuff in setDefaults() to make sure the green select and multifunction LEDs were lit at startup
v1.0.15 - 2015-12-31
	ADD Added serial update on band select
v1.0.14 - 2015-12-31
	CHG Changed serialDump to only send when a value has changed
	CHG Set boundaries for keyer speed (1 to 100WPM, if set to 0 resets to 20)
	CHG Moved lastSerialDump out of global variables into loop() as static
v1.0.13 - 2015-12-31
	ADD Morse code decoder
v1.0.12 - 2015-12-31
	CHG Reorganized multifunction and select button methods to be more compact
	ADD Added reporting and changing of modes via serial commands
	ADD Blink all LEDs at startup
	ADD Keyer on/off serial command
	ADD Initialization finished serial message
v1.0.11 - 2015-12-30
	CHG Set TT LED to light when transmitting
	ADD Added 7 x ditDuration pause when space character is received
	FIX Fixed tuning so the radio actually tunes
	ADD Added set frequency serial command
v1.0.10 - 2015-12-30
	CHG Moved keyer methods to keyer.ino
	ADD Added morse encoder serial command
v1.0.9 - 2015-12-30
	CHG Cleaned up pollRotaryEncoder
	DEL Removed some more unused global variables
	ADD Added incoming serial command support for tune up, tune down, and set keyer wpm
v1.0.8 - 2015-12-30
	ADD Added signal strength to binary serial dump
	DEL Removed a bunch of extra global variables
	CHG Condensed bandwidth and stepsize methods into one
	DEL Removed TimerOverflow stuff (didn't seem to be doing anything)
	DEL Removed TEXTOUTPUT and BINARYOUTPUT; all data will be binary now
v1.0.7 - 2015-12-30
	CHG Tuning methods
v1.0.6 - 2015-12-30
	CHG Renamed global variable frequency to currentFrequency
	CHG Changed multiple bandlimit methods to one
v1.0.5 - 2015-12-30
	ADD Added power in, power out, and signal strength reading methods
	ADD Added binary serial updates
	DEL TX_Routine()
	ADD setTransmit()
	ADD Added iambic keyer
v1.0.4 - 2015-12-30
	CHG Changed global constants into #defines
	ADD Added binary serialSend() methods
	ADD Message type #defines for serialSend()
	ADD TEXTOUTPUT or BINARYOUTPUT #defines to choose serial reporting data type
v1.0.3 - 2015-12-29
	CHG Renamed many functions for consistency and clarity
v1.0.2 - 2015-12-29
	CHG Revised #define statements for consistency
v1.0.1 - 2015-12-29
	CHG Broke the massive single file into logical units
	DEL Removed all LCD stuff
v1.0.0 - 2015-12-29
	Started with code downloaded from http://www.rkrdesignsllc.com/static/media/documents/tt_506__rebel_control_software_v2.0.pde

defines.h
#ifndef DEFINES_H
#define DEFINES_H

#define __NAME__                            "TenTec Rebel 506 (KE4UKZ)"
#define __VERSION__                         "1.0.18"

//PWM on pins 3, 5, 6, 9, & 10

#define PIN_AD9834_SDATA                    10          //  keep!
#define PIN_AD9834_SCLK                     8           //  keep!
#define PIN_AD9834_FSYNC                    9           //  keep!
#define PIN_AD9834_RESET                    11          //  keep!
#define PIN_AD9834_FREQ_REGISTER            12          //  keep!
#define AD9834_FREQ0_REGISTER_SELECT_BIT    0x4000      //  keep!
#define AD9834_FREQ1_REGISTER_SELECT_BIT    0x8000      //  keep!
#define FREQ0_INIT_VALUE                    0x01320000  //  ?
#define REFERENCE                           49.99975e6  // for ad9834 this may be tweaked in software to fine tune the Radio

#define PIN_ENCODER0A                       7           // Rotary encoder A
#define PIN_ENCODER0B                       6           // PWM Rotary encoder B
#define PIN_SIDETONE                        3           // PWM maybe to be changed to a logic control for a separate side tone gen
#define PIN_KEY_DAH                         33          // Physical pin attached to the "dah" input
#define PIN_KEY_DIT                         32          // Physical pin attached to the "dit" input
#define PIN_TRANSMIT                        38          // Physical pin attached to the transmitter key
#define PIN_TT_LED                          24          //
#define PIN_BAND_SELECT                     41          // if shorting block on only one pin 20m(1) on both pins 40m(0)
#define PIN_MULTIFUNCTION_BUTTON            2           // Multifunction button input pin
#define PIN_MULTIFUNCTION_GREEN             34          // For now assigned to BW (Band width)
#define PIN_MULTIFUNCTION_YELLOW            35          // For now assigned to STEP size
#define PIN_MULTIFUNCTION_RED               36          // For now assigned to USER
#define PIN_SELECT_BUTTON                   5           // PWM Select button input pin
#define PIN_SELECT_GREEN                    37          // Wide/100/USER1
#define PIN_SELECT_YELLOW                   39          // Medium/1K/USER2
#define PIN_SELECT_RED                      40          // Narrow/10K/USER3
#define PIN_MEDIUM_A8                       22          // Hardware control of I.F. filter Bandwidth
#define PIN_NARROW_A9                       23          // Hardware control of I.F. filter Bandwidth
#define PIN_CWSPEED                         A7          // To adjust CW speed for user written keyer.
#define PIN_CWREAD                          A6          // Can be used to decode CW. 
#define PIN_POWEROUT                        A3          // Reads RF out voltage at Antenna.
#define PIN_POWERIN                         A2          // Reads 1/5 th or 0.20 of supply voltage.
#define PIN_SMETER                          A1          // To give a realitive signal strength based on AGC voltage.
#define PIN_RIT                             A0          // pin that the sensor is attached to used for a rit routine later.

#define MULTIFUNCTION_1                     0
#define MULTIFUNCTION_2                     1
#define MULTIFUNCTION_3                     2

#define FUNCTION_1                          0
#define FUNCTION_2                          1
#define FUNCTION_3                          2

#define BANDWIDTH_WIDE                      0           // About 2.1 KHZ
#define BANDWIDTH_MEDIUM                    1           // About 1.7 KHZ
#define BANDWIDTH_NARROW                    2           // About 1 KHZ

#define STEP_100HZ                          0           //  100Hz tuning step
#define STEP_1000HZ                         1           // 1000Hz tuning step
#define STEP_10000HZ                        2           //10000Hz tuning step

#define USER_1                              0           //User-defined mode 1
#define USER_2                              1           //User-defined mode 2
#define USER_3                              2           //User-defined mode 3

#define KEYER_DISABLED                      0           //Ignore all keyer input; do not transmit
#define KEYER_STRAIGHT                      1           //Grounding tip transmits as long as it is connected
#define KEYER_IAMBIC                        2           //Grounding tip transmits a dit, grounding ring transmits a dah
#define KEYER_TUNE                          3           //Transmit indefinitely

#define BANDSCAN_SIZE                       256         //Number of elements in the bandscanData array
#define BANDSCAN_OFF                        0
#define BANDSCAN_FULLBAND                   1
#define BANDSCAN_CENTERED                   2
#define BANDSCAN_CUSTOM                     3

#define BAND_20METERS                       0
#define BAND_40METERS                       1
#define BANDLIMIT_20_TOP                    5.35e6
#define BANDLIMIT_20_BOTTOM                 5.0e6
#define BANDLIMIT_40_TOP                    16.3e6
#define BANDLIMIT_40_BOTTOM                 16.0e6
#define DEFAULT_40METERS                    16.03e6     // IF + Band frequency, HI side injection 40 meter range 16 > 16.3 mhz
#define DEFAULT_20METERS                    5.06e6      // Band frequency - IF, LOW side injection 20 meter range 5 > 5.35 mhz

//Data items sent from the Rebel to the computer
#define SERIAL_HEADER                       0xee
#define SERIAL_FOOTER                       0xff
#define SERIAL_STRING                       0x00
#define SERIAL_TXSTART                      0x01
#define SERIAL_TXEND                        0x02
#define SERIAL_VOLTMETER                    0x03
#define SERIAL_TXFREQUENCY                  0x04
#define SERIAL_RXFREQUENCY                  0x05
#define SERIAL_OUTPUTPOWER                  0x06
#define SERIAL_SMETER                       0x07
#define SERIAL_WPM                          0x08
#define SERIAL_RECEIVE_CHAR                 0x09
#define SERIAL_CHAR_SENT                    0x0a
#define SERIAL_BANDWIDTH                    0x0b
#define SERIAL_STEPSIZE                     0x0c
#define SERIAL_USER                         0x0d
#define SERIAL_KEYER                        0x0e
#define SERIAL_INITDONE                     0x0f
#define SERIAL_DECODER                      0x10
#define SERIAL_DECODETHRESHHOLD             0x11
#define SERIAL_BAND                         0x12
#define SERIAL_BANDSCAN                     0x13
#define SERIAL_BANDSCAN_WIDTH               0x14
#define SERIAL_BANDSCAN_TOP                 0x15
#define SERIAL_BANDSCAN_BOTTOM              0x16
#define SERIAL_BANDSCAN_INTERVAL            0x17
#define SERIAL_BANDSCAN_DATA                0x18
#define SERIAL_WRAPAROUND                   0x19
#define SERIAL_CWREAD                       0x1a

//Commands received from the computer
#define SERIAL_SET_FREQUENCY                0x80
#define SERIAL_TUNE_UP                      0x81
#define SERIAL_TUNE_DOWN                    0x82
#define SERIAL_SET_WPM                      0x83
#define SERIAL_SEND_CHAR                    0x84
#define SERIAL_SET_BANDWIDTH                0x85
#define SERIAL_SET_STEPSIZE                 0x86
#define SERIAL_SET_USER                     0x87
#define SERIAL_SET_KEYER                    0x88
#define SERIAL_SET_DECODER                  0x89
#define SERIAL_SET_DECODETHRESHHOLD         0x8a
#define SERIAL_SET_BANDSCAN                 0x8b
#define SERIAL_SET_BANDSCAN_WIDTH           0x8c
#define SERIAL_SET_BANDSCAN_TOP             0x8d
#define SERIAL_SET_BANDSCAN_BOTTOM          0x8e
#define SERIAL_SET_BANDSCAN_INTERVAL        0x8f
#define SERIAL_SET_WRAPAROUND               0x90
#define SERIAL_UPDATE                       0x91

//Steps in processing incoming commands
#define SERIALSTEP_IDLE                     0
#define SERIALSTEP_EXPECT_TYPE              1
#define SERIALSTEP_EXPECT_4MORE             2
#define SERIALSTEP_EXPECT_3MORE             3
#define SERIALSTEP_EXPECT_2MORE             4
#define SERIALSTEP_EXPECT_1MORE             5
#define SERIALSTEP_EXPECT_FOOTER            6

#endif
